### ISSUE TITLE
Panic with pointer to collection field

### ISSUE THREAD

**Issue by sbinq** [mitchellh/copystructure#20]
Pretty artificial case again - don't think this would usually be relevant in production code, just for record:

Seems like pointer to collection always copied as collection itself (pointer is lost). Test case:
```
package main

import (
	"testing"

	"github.com/mitchellh/copystructure"
	"github.com/stretchr/testify/require"
)

func TestCopyPointerToCollection(t *testing.T) {
	input := &[]string{"111"}
	output, err := copystructure.Copy(input)
	require.Nil(t, err)
	require.Equal(t, &[]string{"111"}, output, "received value of type %T instead of %T expected", output, input)
}
```

Output:
```
=== RUN   TestCopyPointerToCollection
--- FAIL: TestCopyPointerToCollection (0.00s)
        Error Trace:    sample_test.go:14
	Error:		Not equal: 
			expected: &[]string{"111"}
			received: []string{"111"}
			
			Diff:
			--- Expected
			+++ Actual
			@@ -1,4 +1,4 @@
			-(*[]string)((len=1) {
			+([]string) (len=1) {
			  (string) (len=3) "111"
			-})
			+}
			 
	Messages:	received value of type []string instead of *[]string expected
		
FAIL
exit status 1
FAIL	_/home/sbinq/work/gopath/src	0.003s
```

This wold lead to panic in the following test case:
```
package main

import (
	"testing"

	"github.com/mitchellh/copystructure"
	"github.com/stretchr/testify/require"
)

func TestCopyStructWithPointerToCollection(t *testing.T) {
	type Struct struct {
		Field *[]string
	}

	input := Struct{
		Field: &[]string{"111"},
	}

	output, err := copystructure.Copy(input)
	require.Nil(t, err)
	require.Equal(t, input, output)
}
```

Output is:
```
=== RUN   TestCopyStructWithPointerToCollection
--- FAIL: TestCopyStructWithPointerToCollection (0.00s)
panic: reflect.Set: value of type []string is not assignable to type *[]string [recovered]
	panic: reflect.Set: value of type []string is not assignable to type *[]string

goroutine 5 [running]:
panic(0x63a8e0, 0xc420013670)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
testing.tRunner.func1(0xc420090180)
	/usr/local/go/src/testing/testing.go:579 +0x25d
panic(0x63a8e0, 0xc420013670)
	/usr/local/go/src/runtime/panic.go:458 +0x243
reflect.Value.assignTo(0x637380, 0xc4200d6d60, 0x97, 0x697382, 0xb, 0x62be80, 0x0, 0x1, 0x1, 0x0)
	/usr/local/go/src/reflect/value.go:2163 +0x35c
reflect.Value.Set(0x62be80, 0xc420032058, 0x196, 0x637380, 0xc4200d6d60, 0x97)
	/usr/local/go/src/reflect/value.go:1333 +0xa4
github.com/mitchellh/copystructure.(*walker).Exit(0xc420098090, 0x7, 0x0, 0x0)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/copystructure/copystructure.go:202 +0x86a
github.com/mitchellh/reflectwalk.walkStruct(0x6567a0, 0xc4200d6cc0, 0x19, 0x687760, 0xc420098090, 0x0, 0x10)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/reflectwalk/reflectwalk.go:330 +0x40c
github.com/mitchellh/reflectwalk.walk(0x6567a0, 0xc4200d6cc0, 0x19, 0x687760, 0xc420098090, 0x0, 0x0)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/reflectwalk/reflectwalk.go:180 +0x4b2
github.com/mitchellh/reflectwalk.Walk(0x6567a0, 0xc4200d6cc0, 0x687760, 0xc420098090, 0x7bd4a0, 0x7fda74688000)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/reflectwalk/reflectwalk.go:84 +0x155
github.com/mitchellh/copystructure.Config.Copy(0x0, 0x0, 0x6567a0, 0xc4200d6cc0, 0x10, 0x63dd60, 0xc42002ff01, 0xc4200135d0)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/copystructure/copystructure.go:69 +0x81
github.com/mitchellh/copystructure.Copy(0x6567a0, 0xc4200d6cc0, 0xf3, 0x30, 0xc42002ff58, 0x45e423)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/copystructure/copystructure.go:13 +0x4c
_/home/sbinq/work/gopath/src.TestCopyStructWithPointerToCollection(0xc420090180)
	/home/sbinq/work/gopath/src/sample_test.go:26 +0x95
testing.tRunner(0xc420090180, 0x6b6db8)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec
exit status 2
FAIL	_/home/sbinq/work/gopath/src	0.005s
```

**Comment by sbinq** [comment]
Hm, actually it turns out this issue might have occured in actual production code, consider the following example (failing exactly the same way as the one above):

```
package main

import (
	"encoding/json"
	"testing"

	"github.com/mitchellh/copystructure"
	"github.com/stretchr/testify/require"
)

func TestCopyStructWithRawJson(t *testing.T) {
	type Struct struct {
		Field *json.RawMessage
	}

	msg := json.RawMessage(`{"foo": "bar"}`)

	input := Struct{
		Field: &msg,
	}

	output, err := copystructure.Copy(input)
	require.Nil(t, err)
	require.Equal(t, input, output)
}
```

Using json.RawMessage instead of *json.RawMessage might be not an option due to the way it is serialized by json package - `{"Field":{"foo":"bar"}}` vs `{"Field":"eyJmb28iOiAiYmFyIn0="}`

**Comment by sbinq** [comment]
Submitted pull-request kind of resolving this issue. Not sure if that's a proper solution though (please see pull request description for notes on this) - but could be at least some starting point for work on solution.

@mitchellh could you please review the mentioned pull request and tell me what you think about accepting/not accepting and the issue in general?

**Comment by sbinq** [comment]
@mitchellh having reviewed current logic more carefully I would suggest a bit different fix for this (more "right" and aligned with existing code I suppose) - please see comments at #24 for details.

#21 should probably be considered obsolete I think (will remove it as soon as any decision on this issue made, keeping it for reference for now).

**Comment by mitchellh** [comment]
Fixed by #24
